CMakeLists.txt

```
@@ -359,6 +361,50 @@ if(BUILD_CLANG_PLUGIN)
       message(SEND_ERROR "CLANG_INCLUDE_PATH ${CLANG_INCLUDE_PATH} does not exist. Please provide clang's internal include path manually: Find the directory where __clang_cuda_runtime_wrapper.h is. Provide this directory for older ROCm versions and the parent directory for newer ones.")
     endif()
     if(WITH_ROCM_BACKEND)
+      # Check if we're building against ROCm LLVM. This code checks if
+      # CLANG_EXECUTABLE_PATH is located within ROCM_PATH.
+      # Note: This might fail in ROCm distributions where ROCm is strongly decomposed
+      # (spack?) but it is more reliable than just parsing the output of clang++ --version.
+      # This is then later used to decide whether we need to make exemptions
+      # from LLVM version compatibility checks.
+      file(GLOB_RECURSE ROCM_FOUND_FILES FOLLOW_SYMLINKS "${ROCM_PATH}/*")
+      if ("${CLANG_EXECUTABLE_PATH}" IN_LIST ROCM_FOUND_FILES)
```

Do you have a case in mind where traversing the full tree is more robust than

```suggestion
      file(REAL_PATH "${CLANG_EXECUTABLE_PATH}" CLANG_REAL_PATH)
      file(REAL_PATH "${ROCM_PATH}" ROCM_REAL_PATH)
      cmake_path(IS_PREFIX ROCM_REAL_PATH "${CLANG_REAL_PATH}" NORMALIZE CLANG_IS_IN_ROCM_PATH)
```

? I'm worried that enumerating all the files in the tree is going to make some network filesystems very unhappy.

---

My understanding is that `cmake_path` is available from cmake 3.20, but the current minimum cmake version is 3.10. Not sure if we want to bump the minimum cmake version because of this?

I can follow your concern. I think we could be more conservative with the paths that we test. Realistically, we would e.g. never expect to find `clang` in the `include` subdirectory (which I imagine might be responsible for a large chunk of the many small files that might be challenging for network filesystems). So perhaps it's enough to query `bin`, `llvm/bin` and `lib/llvm/bin` subdirectories?

Such an approach will probably work for the majority of ROCm installations (especially AMD-packaged ones). The case I'm most concerned about though is ROCm installations where ROCm is split up (spack?), or there is no specific ROCm subdirectory because it's just installed into `/usr`...I suppose in that case, users could always use `-DACPP_EXPERIMENTAL_LLVM=ON`..

---

I've updated the PR to now only scan common `bin` directories (`$ROCM_PATH/bin`, `$ROCM_PATH/llvm/bin`, `$ROCM_PATH/lib/llvm/bin`)